retis/src/collect/collector/ovs/flow_info.rs

```diff
@@ -0,0 +1,324 @@
+//! OpenvSwitch flow information enrichment.
+//!
+//! The OpenvSwitch datapath is made of flows which are comprised of
+//! a match and a list actions. They are uniquely identified by a unique
+//! flow id, or UFID.
+//!
+//! Each of these datapath flows are built as a result of the OpenFlow rule
+//! classification which typically involves many OpenFlow rules. Therefore,
+//! each datapath flow is the result of several OpenFlow rules being matched.
+//!
+//! OpenvSwitch 3.4 supports extracting the OpenFlow rules that contributed to
+//! the creation of each datapath flow through a unixctl command called
+//! "ofproto/detrace".
+//!
+//! This module implements a thread that can query OpenvSwitch for this information
+//! (caching the results) and enrich the event file with this relationship.
+
+use std::collections::HashMap;
+use std::collections::VecDeque;
+use std::sync::mpsc;
+use std::sync::Arc;
+use std::thread;
+use std::time::{Duration, SystemTime};
+
+use anyhow::{anyhow, Result};
+use log::{debug, error, info, warn};
+use ovs_unixctl::OvsUnixCtl;
+
+use crate::core::events::factory::RetisEventsFactory;
+use crate::events::*;
+use crate::helpers::signals::Running;
+
+const MAX_REQUESTS_PER_SEC: u64 = 10;
+const MAX_FLOW_AGE_SECS: u64 = 5;
+
+// A request to enrich a flow
+pub(crate) struct EnrichRequest {
+    ufid: Ufid,
+    flow: u64,
+    sf_acts: u64,
+    ts: SystemTime,
+}
+
+impl EnrichRequest {
+    pub(crate) fn new(ufid: Ufid, flow: u64, sf_acts: u64) -> Self {
+        EnrichRequest {
+            ufid,
+            flow,
+            sf_acts,
+            ts: SystemTime::now(),
+        }
+    }
+}
+
+pub(crate) struct FlowEnricher {
+    // Factory to use for event creation
+    events_factory: Arc<RetisEventsFactory>,
+    // Thread handle
+    thread: Option<thread::JoinHandle<()>>,
+    // Whether ofproto/detrace is supported
+    detrace_supported: bool,
+
+    // Sender and receiver of the channel that is used to request enrichments
+    sender: mpsc::Sender<EnrichRequest>,
+    receiver: Option<mpsc::Receiver<EnrichRequest>>,
+}
+
+impl FlowEnricher {
+    pub(crate) fn new(events_factory: Arc<RetisEventsFactory>) -> Result<Self> {
+        let (sender, receiver) = mpsc::channel::<EnrichRequest>();
```

Comment:
Please use a `sync_channel` instead, to avoid having an infinite growing buffer.

(We use non sync channels in other parts of the code, that's fine in practice for now as their readers are not the bottleneck; I'll fix their use in another PR).